### DISCUSSION TITLE
Field type conversion

### DISCUSSION THREAD

**Discussion by xFrann** [pocketbase/pocketbase#3576] [Q&A] [answered]
Hello,

If for example I have a field that was setup as a number and I need to convert it to string
after the database has been populated with data, is there any way without deleting the existing values?
Is there need for a custom migration script instead?
Thank you

**Comment by ganigeorgiev** [comment, marked as answer]
How did you populate the data in the first place? Is the inserted db data stored in the correct format or you want to convert that too?

If the data is in the correct format (aka. it was inserted via sql script) you can try creating a JS or Go migration that will force change only the collection field type programatically. Here is an example with a [JS migration](https://pocketbase.io/docs/js-migrations/) (_I haven't tested it_):

1. Run `./pocketbase migrate create "change field type"`

2. Replace the generated migration file with something like:
  ```js
  /// <reference path="../pb_data/types.d.ts" />
  migrate((db) => {
    const dao = new Dao(db)
    const collection = dao.findCollectionByNameOrId("yourCollection")

    // find the field and change its type
    // (since it is an "object"/pointer we can modify it in place)
    const field = collection.schema.getFieldByName("yourField")
    field.type = "text"

    // persist the change
    return dao.saveCollection(collection)
  }, /* optional, if you need it you can add a reverse function hre */)
  ```

3. Start the application as usual: `./pocketbase serve`

**Reply by ganigeorgiev** [reply]
Another option could be to create a temp field with the correct type and duplicate your data into it:

```js
migrate((db) => {
  db.newQuery("UPDATE yourCollection SET tempField = CAST(yourField as TEXT)").execute()
}, /* optional, if you need it you can add a reverse function hre */)
```

And then you can delete the old field and rename the temp one back to the original from the Admin UI (_it will create automatically a migration for you_).

**Reply by xFrann** [reply]
Hi,

The data has been added trough the API until now, only numbers added (It's a field accepting 0 to 100)
I want to change that field to string but keeping the value (instead of number format, string format)
The migration script you have provided works perfectly for my use case.
Thank you.

**Reply by ganigeorgiev** [reply]
In this case I recommend using the second approach because the stored values will remain unchanged even though it is returned as string in the APIs (SQLite allows storing mixed types) and this could have side-effects for some queries.
Or atleast make sure that everything for that field is stored as text, aka. create a new migration with something like:

```js
migrate((db) => {
  // ensures that the stored column values are text
  db.newQuery("UPDATE yourCollection SET yourField = CAST(yourField as TEXT)").execute()
})
```

**Reply by xFrann** [reply]
You're awesome, thank you
